clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java

```diff
@@ -0,0 +1,454 @@
+        long raw = readRawVarUnsignedLong(in);
+        // This undoes the trick in writeSignedVarLong()
+        long temp = (((raw << 63) >> 63) ^ raw) >> 1;
+        // This extra step lets us deal with the largest signed values by treating
+        // negative results from read unsigned methods as like unsigned values
+        // Must re-flip the top bit if the original read value had it set.
+        return temp ^ (raw & (1L << 63));
+    }
+
+    /**
+     * Read a long stored in variable-length format using Zig-zag decoding from
+     * <a href="http://code.google.com/apis/protocolbuffers/docs/encoding.html">
+     * Google Protocol Buffers</a>. Also update the index to indicate how many bytes
+     * were used to encode this long.
+     *
+     * @param buffer The buffer to read from
+     * @return The integer read (MUST BE TREATED WITH SPECIAL CARE TO AVOID SIGNEDNESS)
+     *
+     * @throws IllegalArgumentException if variable-length value does not terminate
+     *                                  after 5 bytes have been read
+     */
+    public static long readVarLong(ByteBuffer buffer)  {
+        long raw = readRawVarUnsignedLong(buffer);
+        // This undoes the trick in writeSignedVarLong()
+        long temp = (((raw << 63) >> 63) ^ raw) >> 1;
+        // This extra step lets us deal with the largest signed values by treating
+        // negative results from read unsigned methods as like unsigned values
+        // Must re-flip the top bit if the original read value had it set.
+        return temp ^ (raw & (1L << 63));
+    }
+
+    /**
+     * Write the given unsigned integer following the variable-length from
+     * <a href="http://code.google.com/apis/protocolbuffers/docs/encoding.html">
+     * Google Protocol Buffers</a> into the output. Since it the value is not negative Zig-zag is not used.
+     *
+     * @param longValue The value to write
+     * @param out The output to write to
+     */
+    public static void writeVarUnsignedInt(long longValue, DataOutput out) throws IOException {
+        int value = (int) longValue;
+        while ((value & 0xffffff80) != 0L) {
+            out.writeByte((value & 0x7f) | 0x80);
+            value >>>= 7;
+        }
+        out.writeByte(value & 0x7f);
+    }
+
+    /**
+     * Write the given unsigned integer following the variable-length from
+     * <a href="http://code.google.com/apis/protocolbuffers/docs/encoding.html">
+     * Google Protocol Buffers</a> into the output. Since the value is not negative, Zig-zag is not used.
+     *
+     * @param longValue The value to write
+     * @param buffer The buffer to write to
+     */
+    public static void writeVarUnsignedInt(long longValue, ByteBuffer buffer) {
+        int value = (int) longValue;
+        while ((value & 0xffffff80) != 0L) {
+            byte b = (byte) ((value & 0x7f) | 0x80);
+            buffer.put(b);
+            value >>>= 7;
+        }
+        buffer.put((byte) (value & 0x7f));
```

Comment:
I think you are right. Breaking from the loop ensures that the upper nibble of the last byte is already 0, so the `& 0x7f` should be redundant.